File: android/sdk/src/main/java/com/viant/forgeandroid/runtime/WindowFormRuntime.kt
package com.viant.forgeandroid.runtime

private const val WINDOW_FORM_META_KEY = "__forge"
private const val WINDOW_FORM_PREFILL_KEY = "prefill"
private const val WINDOW_FORM_PREFILL_REVISION_KEY = "prefillRevision"
private const val WINDOW_FORM_REPLACE_KEY = "\$replace"
private const val WINDOW_FORM_REPLACE_VALUE_KEY = "value"

private data class WindowFormReplacement(val value: Any?)

internal fun ForgeRuntime.windowFormValue(windowId: String): Map<String, Any?> {
    return windowContext(windowId).peekWindowForm()
}

internal fun ForgeRuntime.setWindowFormValue(
    windowId: String,
    values: Map<String, Any?>,
    replace: Boolean = false,
    bumpPrefillRevision: Boolean = true
) {
    val dsId = WindowIdentity(windowId).windowFormId()
    val signal = windowContext(windowId).signals.form(dsId)
    val nextValues = if (bumpPrefillRevision) {
        withWindowFormPrefillRevision(signal.peek(), values)
    } else {
        values
    }
    if (replace) {
        signal.set(nextValues)
        return
    }
    signal.set(mergeWindowFormValues(signal.peek(), nextValues))
}

internal fun ForgeRuntime.reconcileWindowForm(
    windowId: String,
    metadata: WindowMetadata,
    parameters: Map<String, Any?>
) {
    val seeded = mergeWindowFormValues(parameters, resolveInitialWindowFormValues(metadata))
    val current = windowFormValue(windowId)
    setWindowFormValue(windowId, mergeWindowFormValues(seeded, current), replace = true, bumpPrefillRevision = false)
}

private fun resolveInitialWindowFormValues(metadata: WindowMetadata): Map<String, Any?> {
    val initial = linkedMapOf<String, Any?>()
    val entries = buildList {
        addAll(metadata.on)
        addAll(metadata.window?.on.orEmpty())
    }
    entries
        .filter { it.event == "onInit" && it.handler == "dataSource.setWindowFormData" }
        .forEach { execution ->
            execution.parameters
                .filter { it.input == "const" }
                .forEach { parameter ->
                    val name = parameter.name?.trim().orEmpty()
                    if (name.isNotBlank()) {
                        initial[name] = parameter.locationAny() ?: parameter.value?.let(JsonUtil::elementToAny)
                    }
                }
        }
    metadata.view?.content?.containers.orEmpty().forEach { container ->
        collectInitialWindowFormItemValues(container, initial)
    }
    return initial
}

private fun collectInitialWindowFormItemValues(
    container: ContainerDef,
    initial: MutableMap<String, Any?>
) {
    container.items
        .filter { it.scope?.trim() == "windowForm" }
        .forEach { item ->
            val key = item.valueKey().orEmpty()
            val value = item.value?.let(JsonUtil::elementToAny)
            if (key.isBlank() || value == null) return@forEach
            initial[key] = value
        }
    container.containers.forEach { child ->
        collectInitialWindowFormItemValues(child, initial)
    }
}

private fun mergeWindowFormValues(
    base: Map<String, Any?>,
    override: Map<String, Any?>
): Map<String, Any?> {
    val result = base.toMutableMap()
    override.forEach { (key, value) ->
        val replacement = windowFormReplacementValue(value)
        if (replacement != null) {
            result[key] = replacement.value
            return@forEach
        }
        val current = JsonUtil.asStringMap(result[key])
        val next = JsonUtil.asStringMap(value)
        result[key] = if (current.isNotEmpty() && next.isNotEmpty()) {
            mergeWindowFormValues(current, next)
        } else {
            value
        }
    }
    return result
}

private fun windowFormReplacementValue(value: Any?): WindowFormReplacement? {
    val objectValue = JsonUtil.asStringMap(value)
    if (objectValue[WINDOW_FORM_REPLACE_KEY] != true) {
        return null
    }
    return WindowFormReplacement(objectValue[WINDOW_FORM_REPLACE_VALUE_KEY])
}

private fun withWindowFormPrefillRevision(
    previous: Map<String, Any?>,
    values: Map<String, Any?>
): Map<String, Any?> {
    if (!values.containsKey(WINDOW_FORM_PREFILL_KEY)) {
        return values
    }
    val previousMeta = JsonUtil.asStringMap(previous[WINDOW_FORM_META_KEY])
    val nextMeta = JsonUtil.asStringMap(values[WINDOW_FORM_META_KEY])
    val previousRevision = numericValue(previousMeta[WINDOW_FORM_PREFILL_REVISION_KEY]) ?: 0
    val next = values.toMutableMap()
    next[WINDOW_FORM_META_KEY] = nextMeta + (WINDOW_FORM_PREFILL_REVISION_KEY to previousRevision + 1)
    return next
}

private fun numericValue(value: Any?): Int? {
    return when (value) {
        is Number -> value.toInt()
        is String -> value.trim().toIntOrNull()
        else -> null
    }
}

File: ios/Sources/ForgeIOSRuntime/WindowFormRuntime.swift
import Foundation

extension ForgeRuntime {
    public func windowFormJSONValue(windowID: String) async -> [String: JSONValue] {
        let dataSourceID = WindowIdentity(windowID: windowID).windowFormID()
        return await dataSourceRuntime.form(dataSourceID: dataSourceID)
    }

    public func setWindowFormValue(
        windowID: String,
        values: [String: JSONValue],
        replace: Bool = false,
        bumpPrefillRevision: Bool = true
    ) async {
        let dataSourceID = WindowIdentity(windowID: windowID).windowFormID()
        let signal = await signals.form(dataSourceID: dataSourceID)
        let current = await dataSourceRuntime.form(dataSourceID: dataSourceID)
        let nextValues = bumpPrefillRevision
            ? valuesWithPrefillRevision(values, current: current)
            : values
        if replace {
            await dataSourceRuntime.setForm(dataSourceID: dataSourceID, values: nextValues)
            await signal.set(nextValues)
            return
        }
        let merged = mergeWindowFormValues(base: current, override: nextValues)
        await dataSourceRuntime.setForm(
            dataSourceID: dataSourceID,
            values: merged
        )
        await signal.set(merged)
    }

    func reconcileWindowForm(
        windowID: String,
        metadata: WindowMetadata,
        parameters: [String: JSONValue]
    ) async {
        let initialValues = resolveInitialWindowFormValues(metadata: metadata)
        let seeded = mergeWindowFormValues(base: parameters, override: initialValues)
        let current = await windowFormJSONValue(windowID: windowID)
        let reconciled = mergeWindowFormValues(base: seeded, override: current)
        await setWindowFormValue(
            windowID: windowID,
            values: reconciled,
            replace: true,
            bumpPrefillRevision: false
        )
    }
}

private func valuesWithPrefillRevision(
    _ values: [String: JSONValue],
    current: [String: JSONValue]
) -> [String: JSONValue] {
    guard values["prefill"] != nil else {
        return values
    }
    var next = values
    let currentMeta = current["__forge"]?.objectValue ?? [:]
    let incomingMeta = values["__forge"]?.objectValue ?? [:]
    var meta = mergeWindowFormValues(base: currentMeta, override: incomingMeta)
    meta["prefillRevision"] = .number(Double(prefillRevision(from: currentMeta) + 1))
    next["__forge"] = .object(meta)
    return next
}

private func prefillRevision(from values: [String: JSONValue]) -> Int {
    switch values["prefillRevision"] {
    case .number(let value):
        return Int(value)
    case .string(let value):
        return Int(value.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
    default:
        return 0
    }
}

private func resolveInitialWindowFormValues(metadata: WindowMetadata) -> [String: JSONValue] {
    var initial: [String: JSONValue] = [:]
    let entries = metadata.on
    for entry in entries where entry.event == "onInit" && entry.handler == "dataSource.setWindowFormData" {
        for parameter in entry.parameters where parameter.input == "const" {
            let name = parameter.name.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
            guard !name.isEmpty else { continue }
            if let location = parameter.location {
                initial[name] = location
            } else if let value = parameter.value {
                initial[name] = value
            }
        }
    }
    for container in metadata.view?.content?.containers ?? [] {
        collectInitialWindowFormItemValues(container: container, initial: &initial)
    }
    return initial
}

private func collectInitialWindowFormItemValues(
    container: ContainerDef,
    initial: inout [String: JSONValue]
) {
    for item in container.items where (item.scope ?? "").trimmingCharacters(in: .whitespacesAndNewlines) == "windowForm" {
        let fieldKey = (item.bindingPath ?? item.dataField ?? item.id ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !fieldKey.isEmpty, let value = item.value else { continue }
        initial[fieldKey] = value
    }
    for child in container.containers {
        collectInitialWindowFormItemValues(container: child, initial: &initial)
    }
}

private func mergeWindowFormValues(
    base: [String: JSONValue],
    override: [String: JSONValue]
) -> [String: JSONValue] {
    var result = base
    for (key, value) in override {
        if
            case .object(let currentObject)? = result[key],
            case .object(let overrideObject) = value
        {
            result[key] = .object(mergeWindowFormValues(base: currentObject, override: overrideObject))
        } else {
            result[key] = value
        }
    }
    return result
}

extension JSONValue {
    init?(any: Any?) {
        guard let any else {
            self = .null
            return
        }
        switch any {
        case let value as JSONValue:
            self = value
        case let value as String:
            self = .string(value)
        case let value as Bool:
            self = .bool(value)
        case let value as Int:
            self = .number(Double(value))
        case let value as Int64:
            self = .number(Double(value))
        case let value as Double:
            self = .number(value)
        case let value as Float:
            self = .number(Double(value))
        case let value as NSNumber:
            self = .number(value.doubleValue)
        case let value as [String: JSONValue]:
            self = .object(value)
        case let value as [String: Any]:
            var object: [String: JSONValue] = [:]
            for (key, child) in value {
                guard let jsonValue = JSONValue(any: child) else { return nil }
                object[key] = jsonValue
            }
            self = .object(object)
        case let value as [JSONValue]:
            self = .array(value)
        case let value as [Any]:
            let array = value.compactMap(JSONValue.init(any:))
            guard array.count == value.count else { return nil }
            self = .array(array)
        default:
            return nil
        }
    }

    var anyValue: Any? {
        switch self {
        case .string(let value):
            return value
        case .number(let value):
            return value
        case .bool(let value):
            return value
        case .array(let value):
            return value.map(\.anyValue)
        case .object(let value):
            return value.mapValues(\.anyValue)
        case .null:
            return nil
        }
    }
}
